src/v/kafka/server/group_manager.cc

```diff
@@ -746,6 +762,75 @@ ss::future<> group_manager::do_recover_group(
     co_return;
 }
 
+ss::future<> group_manager::write_version_fence(
+  model::term_id term, ss::lw_shared_ptr<attached_partition> p) {
+    // how long to delay retrying a fence write if an error occurs
+    constexpr auto fence_write_retry_delay = 10s;
+
+    co_await _feature_table.local().await_feature(
+      features::feature::group_offset_retention, p->as);
+
+    while (true) {
+        if (p->as.abort_requested() || _gate.is_closed()) {
+            break;
+        }
+
+        auto batch = _feature_table.local().encode_version_fence(
+          cluster::cluster_version{9});
+        auto reader = model::make_memory_record_batch_reader(std::move(batch));
+
+        try {
+            auto result = co_await p->partition->raft()->replicate(
+              term,
+              std::move(reader),
+              raft::replicate_options(raft::consistency_level::quorum_ack));
+
+            if (result) {
+                vlog(
+                  klog.info,
+                  "Prepared partition {} for consumer offset retention feature "
+                  "during upgrade",
+                  p->partition->ntp());
+                break;
+
+            } else if (result.error() == raft::errc::shutting_down) {
+                vlog(
+                  klog.debug,
+                  "Cannot write offset retention version fence for partition "
+                  "{}: shutting down",
+                  p->partition->ntp());
+                break;
+
+            } else if (result.error() == raft::errc::not_leader) {
+                vlog(
+                  klog.debug,
+                  "Cannot write offset retention version fence for partition "
+                  "{}: not leader",
+                  p->partition->ntp());
+                break;
```

Comment:
nit: here and elsewhere `co_return` better signals the intent to readers IMO (especially since this is a kinda tall method)

Reply:
@andrwng thanks I posted a PR addressing your nits: https://github.com/redpanda-data/redpanda/pull/8660